Issue 1669: Use new BitMask struct for validators

hewison-chris opened the pull request:
This is the field that stores the information of which validators have signed the block. The new field type will enable us to set the count when we construct the field and include it in the BlockHeader hash.

hewison-chris committed:
`BitMask` `struct` for `validators`

The `BitMask` `struct` is specialized for use as the field in the
`BlockHeader` to record which validators have signed that block.
The number of active validators allowed to sign the block are set in the
constructor and can not be changed.
Trying to read or write to bits beyond the count is not allowed and will
cause an assertion.

Geod24 commented on source/agora/common/BitField.d:
```diff
@@ -27,8 +27,8 @@
 
-module agora.common.BitField;
 
+module agora.common.BitField;
```
?

hewison-chris replied:
Reverted

linked0 commented:
LGTM. But it failed on CI tests.

hewison-chris commented:
> LGTM. But it failed on CI tests.

Will fix the tests now.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,195 @@
+    /// String serialization
+    public string toString () const
+    {
+        return iota(this.count).map!(i => this[i] ? "1" : "0").join("");
```
As discussed IRL: `"4/6 (2, 4)"`, where the number between parenthesis is 0-indexed.
Also needs to use a sink.

hewison-chris replied:
Will update. Just getting the tests to pass first.

hewison-chris replied:
This will be done for PrettyPrint only as for vibe.d we will transfer as string of bit values (e.g. `01001111`)

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,195 @@
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
+    }
+
+    // set the bits that are set in given BitMask
+    public auto addBitMask (in BitMask add)
+    {
+        assert(this.count == add.count, "BitMask assignment must be for same bit count");
+        iota(this.count).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
+        return this;
+    }
```
That's `|=`, aka `opOpAssign!"|"`
https://dlang.org/spec/operatoroverloading.html#op-assign

hewison-chris replied:
TIL

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,195 @@
+
+    @nogc:
+
+    /// return the count of set bits
+    public size_t setCount () const
+    {
+        return iota(this.count).count!(i => this[i]);
+    }
```
`length` is the standard name for this.

hewison-chris replied:
Updated to length

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,195 @@
+        return (1 << (8 - 1 - (index % 8)));
+    }
+
+    /// Support for sorting by count of set bits not value
+    public int opCmp (in typeof(this) rhs) const
+    {
+        assert(this.count == rhs.count, "Comparing different sized BitMasks is not valid");
+        return this.setCount < rhs.setCount ? -1 : 1;
+    }
```
Also need `opEquals`

hewison-chris replied:
I think the default for this struct with length first and then `ubyte` array second is ok. Or maybe I am missing something?

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,195 @@
+
+module agora.common.BitMask;
+
+import agora.serialization.Serializer;
```
Only used in UT ?

hewison-chris replied:
It is now. Will make import for `version unittest`.

hewison-chris commented:
Updated based on review. Will fix any tests that might fail in the morning.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,230 @@
+    {
+        import std.format;
+        bool list_set = this.setCount <= this.length / 2;
+        formattedWrite(sink, "%d/%d %s", this.setCount, this.length,
```
```suggestion
        formattedWrite(sink, "%d/%d %(%s, %)", this.setCount, this.length,
```

And then just give it a range.

hewison-chris replied:
What about the `!`?

Geod24 replied:
I say, let's just list all missing validators, even if the list is longer.
This is for human consumption, I don't think we should use this for Vibe.d serialization. I'll have a look at what is available.

hewison-chris replied:
But the idea is to reduce the bandwidth by always using the shortest list.

Geod24 replied:
What bandwidth ?

hewison-chris replied:
If we have 1000 validators and we list 999 indices yet to sign we use more bandwidth than listing the first one that signs.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,230 @@
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto set = str.findSplitBefore("/")[0].to!size_t;
+        auto length = str.findSplitAfter("/")[1].findSplitBefore(" ")[0].to!size_t;
+        bool list_set = set <= length / 2;
+        auto indices = str.findSplitAfter(list_set ? "(" : "!(")[1].findSplitBefore(")")[0]
+            .split(",").map!(i => i.to!size_t);
+        if (list_set)
+            assert(indices.length == set);
+        else
+            assert(set + indices.length == length);
+        auto bitmask = BitMask(length);
+        iota(length).each!((size_t i)
+        {
+            if (indices.canFind(i) == list_set)
+                bitmask[i] = true;
+        });
+        return bitmask;
```
Ah this affects how Vibe.d serializes it right ?

hewison-chris replied:
Now the `fromString` reads string of bit values (e.g. `0010101111`)

hewison-chris commented:
Green again.

linked0 reviewed (approved):
LGTM

ferencdg commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,192 @@
+        return this.setCount < rhs.setCount ? -1 : 1;
+    }
+
+    @nogc:
```
can we explicitely declare @nogc/@safe/nothrow on the functions instead? sometimes we re-arrange functions inside a file, and now the function's signature depends on where the function appears in the file

hewison-chris replied:
I was following a style I had seen used in some other PRs recently. I am really not sure which is better. Position the functions based on attributes or explicitly specify them on each function. Any other opinions on this?

ferencdg replied:
hmm, I remember I used to see some Dlang code (outside our codebase), where I had to scroll through an entire file just to be able to see what the modifiers on a functions are

Geod24 replied:
Yeah I used it in Bitblob in the following case:
```D
struct BitBlob (size_t byteSize)
{
    @safe pure: // Everything
    this () {} // Constructor that `throw new`, so not `nothrow @nogc`
    string toString() {} // Allocates, obviously
    void toString (SinkType sink) {} // Does not allocate, but not provable. And can `throw new`
    nothrow @nogc: // Everything else
}
```

In other word, when the number of methods that don't get the attributes applied is very small, and we want the attribute to apply to everything, I sometimes use `attribute:`. But in this case (`@nogc`), it'd be better to have it on the methods for sure. The other attributes, I'm neutral.

hewison-chris replied:
As `@safe` applies to all methods, `pure` all but two and `nothrow` all but three I think it makes sense to keep them as is. I'll make the `@nogc` explicit as  it only applies to three methods.

MichaelKim20 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,192 @@
+        if (bit_index >= this.length)
+                assert(0, "Attempt to set index beyond length of bitmask");
+        const size_t byte_index = (bit_index) / 8;
+        const size_t bit_index_in_byte = bit_index % 8;
```
Operation % is already implemented in function mask().

hewison-chris replied:
Good catch, I missed that.

Geod24 commented:
Just curious, what's the story behind the first commit ? It doesn't seem related AFAICS.

hewison-chris commented:
> Just curious, what's the story behind the first commit ? It doesn't seem related AFAICS.

It seemed to always fail on that test without the update to which nodes we check. It may not be needed anymore following other fixes that have been made since. I'll try again without it.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,189 @@
+        return iota(this.length).filter!(i => !this[i]);
+    }
+
+    /// return the length of set bits
```
```suggestion
    /// Returns: the number of bits set to `true`
```

Reading the comment I thought it was the `length`

hewison-chris replied:
Updated

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,189 @@
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t length;
```
This should be `private`.

hewison-chris replied:
Updated and added public `count` function.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,189 @@
+    public this (size_t length, in ubyte[] bytes)
+    {
+        this(length);
+        assert(this.bytes.length == bytes.length);
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
```
```suggestion
        this.bytes[] = bytes;
```
It will throw a `RangeError` if lengths don't match.

hewison-chris replied:
Updated

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -0,0 +1,189 @@
+    public auto opOpAssign (string op : "|") (in BitMask add)
+    {
+        assert(this.length == add.length, "BitMask assignment must be for same bit length");
+        iota(this.length).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
```
```suggestion
        this.bytes |= add.bytes;
```

Should work.

Geod24 replied:
Note that you can do the same with `&=` if you ever want to have the ability to merge two sets.

hewison-chris replied:
Updated but needed to `slice` the arrays.

hewison-chris commented:
> > Just curious, what's the story behind the first commit ? It doesn't seem related AFAICS.
> 
> It seemed to always fail on that test without the update to which nodes we check. It may not be needed anymore following other fixes that have been made since. I'll try again without it.

Test passes without the first commit so have removed.
@Geod24 - can we merge?

Geod24 commented on source/agora/consensus/data/genesis/Coinnet.d:
```diff
@@ -34,7 +34,7 @@ public immutable Block GenesisBlock =
         merkle_root: GenesisMerkleRoot,
         enrollments: Enrollments,
-        validators:  BitField!ubyte(6),
+        validators:  BitMask(0),
```
Should probably still be 6

hewison-chris replied:
But they are not `active` until block 1. They just have enrolments in Genesis block and do not sign.

Geod24 replied:
Good point! Ignore then. Or perhaps add your explanation as a comment ?

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -1035,6 +1035,7 @@ public class NetworkManager
                             {
                                 try
                                 {
+                                    log.trace("getMissingBlockSigs: updating header signature: {} validators: {}", header.signature, header.validators);
```
Won't the nominator log this ? In any case, unrelated to the commit.

hewison-chris replied:
This is logged during catch up and not during nomination. I'll split it into a different commit.

Geod24 commented on source/agora/test/PeriodicCatchup.d:
```diff
@@ -63,6 +59,12 @@ private extern(C++) class DoesNotExternalizeBlockNominator : TestNominator
             assert(0, format!"PeriodicCatchup exception thrown during test: %s"(e));
         }
     }
+
+    extern(D) public override void receiveBlockSignature (in ValidatorBlockSig block_sig) @safe
+    {
+        if (block_sig.height == 1)
+            log.trace("Ignore signatures for block 1 to test signature catchup");
+    }
```
Why this change ?

hewison-chris replied:
This adds a test for missing signatures to the `PeriodicCatchup` Network test. I'll extract it to a separate commit with the trace log line above (which triggered me to update this test).

hewison-chris replied:
In fact this was a required change as the validators can only be added (once a bit is set it can not be unset).

Geod24 reviewed (approved):
Nice, just a couple nits.

hewison-chris committed:
Unit test PeriodicCatchup updated

This change is to test signature catchup without removing the validator
signatures from the header. In the next commit it will no longer be
possible to remove signatures (i.e. once a validator has signed a block
it is not possible to remove that signature).
